### PR TITLE
Explore: Replaces TimeSeries with GraphSeriesXY

### LINKED ISSUE #17302: Replace TimeSeries in Explore and Graph component in Explore to use GraphSeriesXY instead
We're moving toward using `GraphSeriesXY` instead of TimeSeries. As part of this we should remove the usage of TimeSeries in Explore and the Graph in Explore.
https://github.com/grafana/grafana/blob/master/public/app/plugins/panel/graph2/getGraphSeriesModel.ts#L30

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
To consolidate the number of Graph components in Grafana this PR makes Explore use the React Graph component from grafana/ui. This means that TimeSeries is no longer used by Explore.

**Which issue(s) this PR fixes**:
Fixes #17302

**Special notes for your reviewer**:

### CODE REVIEW

File: packages/grafana-data/src/types/graph.ts [hunk 1]
```diff
@@ -1,5 +1,11 @@
 import { DisplayValue } from './displayValue';
 
+export interface YAxis {
```
Comment (hugohaggmark) [inline packages/grafana-data/src/types/graph.ts]: This is as I interpreted our internal discussion but not sure that it's ok.

File: packages/grafana-data/src/utils/processDataFrame.ts [hunk 2]
```diff
@@ -44,6 +44,23 @@ function convertTimeSeriesToDataFrame(timeSeries: TimeSeries): DataFrame {
   };
 }
 
+function convertGraphSeriesToDataFrame(graphSeries: GraphSeriesXY): DataFrame {
```
Comment (hugohaggmark) [inline packages/grafana-data/src/utils/processDataFrame.ts]: Not really sure about this one.

File: packages/grafana-ui/src/types/panel.ts [hunk 3]
```diff
@@ -27,6 +27,7 @@ export interface PanelProps<T = any> {
   // TODO: annotation?: PanelData;
 
   timeRange: TimeRange;
+  timeZone: TimeZone;
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/types/panel.ts]: I added TimeZone here as it is needed in the Graph to display correct values on axis. Not sure this is the correct way.

File: packages/grafana-ui/src/utils/flotPairs.ts [hunk 4]
```diff
@@ -1,16 +1,14 @@
 // Types
-import { NullValueMode, DataFrame, GraphSeriesValue } from '@grafana/data';
+import { NullValueMode, GraphSeriesValue } from '@grafana/data';
 
 export interface FlotPairsOptions {
-  series: DataFrame;
+  rows: any[][];
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/utils/flotPairs.ts]: Changed getFlotPairs so it takes rows instead of series because we internally only used rows anyway. Didn't want to create a DataFrame from already parsed GraphSeriesXY

File: public/app/features/dashboard/dashgrid/PanelChrome.tsx [hunk 5]
```diff
@@ -178,7 +178,7 @@ export class PanelChrome extends PureComponent<Props, State> {
         queries: panel.targets,
         panelId: panel.id,
         dashboardId: this.props.dashboard.id,
-        timezone: this.props.dashboard.timezone,
+        timezone: this.props.dashboard.getTimezone(),
```
Comment (hugohaggmark) [inline public/app/features/dashboard/dashgrid/PanelChrome.tsx]: `this.props.dashboard.timezone` was always an empty string, could be a bug. But to get this working I called `this.props.dashboard.getTimezone()` instead.

File: public/app/features/explore/Explore.tsx [hunk 6]
```diff
@@ -192,7 +193,7 @@ export class Explore extends React.PureComponent<ExploreProps> {
   refreshExplore = () => {
     const { exploreId, update } = this.props;
 
-    if (update.queries || update.ui || update.range || update.datasource) {
+    if (update.queries || update.ui || update.range || update.datasource || update.mode) {
```
Comment (hugohaggmark) [inline public/app/features/explore/Explore.tsx]: Changed so that we use value from state before url because of strange behaviour when using the Metrics/Logs switch.

File: public/app/features/explore/Explore.tsx [hunk 7]
```diff
@@ -318,15 +323,15 @@ function mapStateToProps(state: StoreState, { exploreId }: ExploreProps) {
     const urlModeIsValid = supportedModes.includes(urlMode);
     const modeStateIsValid = supportedModes.includes(mode);
 
-    if (urlModeIsValid) {
-      newMode = urlMode;
-    } else if (modeStateIsValid) {
+    if (modeStateIsValid) {
```
Comment (hugohaggmark) [inline public/app/features/explore/Explore.tsx]: Changed so that we use value from state before url because of strange behaviour when using the Metrics/Logs switch.

File: public/app/features/explore/ExploreGraphPanel.tsx [hunk 8]
```diff
@@ -0,0 +1,174 @@
+import React, { PureComponent } from 'react';
```
Comment (hugohaggmark) [inline public/app/features/explore/ExploreGraphPanel.tsx]: This is the Explore version of GraphPanel, don't agree totally with the name. Any suggestions very appreciated.

File: public/app/features/explore/QueryRow.tsx [hunk 9]
```diff
@@ -217,7 +217,7 @@ function mapStateToProps(state: StoreState, { exploreId, index }: QueryRowProps)
   const query = queries[index];
   const datasourceStatus = datasourceError ? DataSourceStatus.Disconnected : DataSourceStatus.Connected;
   const error = queryErrors.filter(queryError => queryError.refId === query.refId)[0];
-  const series = graphResult ? graphResult : []; // TODO: use DataFrame
+  const series = graphResult ? graphResult.map(serie => guessFieldTypes(toDataFrame(serie))) : []; // TODO: use DataFrame
```
Comment (hugohaggmark) [inline public/app/features/explore/QueryRow.tsx]: Not sure why QueryRows need the series.

File: public/app/plugins/panel/graph2/GraphSeriesToggler.tsx [hunk 10]
```diff
@@ -0,0 +1,85 @@
+import React from 'react';
```
Comment (hugohaggmark) [inline public/app/plugins/panel/graph2/GraphSeriesToggler.tsx]: We can't use GraphController in Explore because it needs DataFrame[] as data. The GrapSeriesToggler uses GraphSeriesXY instead and can be used by both GraphController and ExploreGraphPanel. Not sure if this component is the correct way to solve this. Not sure about the name as well.

File: packages/grafana-ui/src/components/Graph/Graph.tsx [hunk 11]
```diff
@@ -4,41 +4,66 @@ import React, { PureComponent } from 'react';
...
+      this.$element = $(this.element);
+      this.$element.bind('plotselected', this.onPlotSelected);
+    }
+  }
+
+  componentWillUnmount() {
+    this.$element.unbind('plotselected', this.onPlotSelected);
   }
 
+  onPlotSelected = (event: JQueryEventObject, ranges: { xaxis: { from: number; to: number } }) => {
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/components/Graph/Graph.tsx]: This is only working from Explore as of now. Needs some work how to solve this in the "Panel" case. Use TimeSrv directly or Redux. Leaving this for a later PR to solve #18476.

File: packages/grafana-ui/src/types/panel.ts (hunk 3 above)
Comment (torkelo) [reply packages/grafana-ui/src/types/panel.ts]: think your right that this is needed

File: public/app/features/explore/QueryRow.tsx (hunk 9 above)
Comment (hugohaggmark) [reply public/app/features/explore/QueryRow.tsx]: Used by for instance PromQueryField for hinting.

**Review by davkal — COMMENTED** [review]
I have been testing quite a bit with Prometheus, Loki, Elastic. Could not find any issues.
Code changes and file naming looks good too.

File: packages/grafana-data/src/types/graph.ts [hunk 12]
```diff
@@ -1,5 +1,11 @@
...
+  index: number;
```
Comment (ryantxu) [inline packages/grafana-data/src/types/graph.ts]: I am a bit confused by the index here -- I figured it was always 1.  I think maybe @dprokop added it to support multiple y axis?  

I think that is unrelated to the proposed changes though.

**Review by ryantxu — APPROVED** [review]
This looks like a big improvement -- good work.

I am hoping that DataFrame could directly support the needs of LogsModel, but it is not ready for that yet.  In particular, the `series` is a histogram with visualisation properties attached (color from log level).  Ideally the histogram could come from a query (elasticsearch) or be computed (loki) and the visualization properties added in ExploreGraphPanel.

File: packages/grafana-data/src/utils/processDataFrame.ts (hunk 2 above)
Comment (hugohaggmark) [reply packages/grafana-data/src/utils/processDataFrame.ts]: Good point, seems like we even revert it back to legacydata:
https://github.com/grafana/grafana/blob/8b0bbb2c4f316fe5140f585100fb5f268a5532f6/public/app/plugins/datasource/prometheus/components/PromQueryField.tsx#L178

I'll see if I can get rid of this.

Comment (hugohaggmark) [reply packages/grafana-data/src/utils/processDataFrame.ts]: I've looked into this but I can't find an easy fix for it. I think we've to introduce PanelData for the Explore state and that seems to be a bigger PR so I'll keep this for now.
The problem is this line https://github.com/grafana/grafana/blob/b3a75cee680d90dd51cd8bc41ebade51a10636f5/public/app/features/explore/QueryRow.tsx#L220
that needs to be solved.

File: public/app/features/dashboard/dashgrid/PanelChrome.tsx (hunk 5 above)
Comment (hugohaggmark) [reply public/app/features/dashboard/dashgrid/PanelChrome.tsx]: @torkelo @ryantxu any thoughts on this one?

File: packages/grafana-data/src/utils/processDataFrame.ts (hunk 2 above)
Comment (torkelo) [reply packages/grafana-data/src/utils/processDataFrame.ts]: yes, think we need to add the raw DataFrame result to the ExploreState as going from GraphXYSeries back to DataFrame can't be good :)

File: packages/grafana-ui/src/components/Graph/Graph.tsx (hunk 11 above)
Comment (torkelo) [reply packages/grafana-ui/src/components/Graph/Graph.tsx]: We will need to add OnTimeRangeChange to PanelProps (in other PR unrelated to this)

File: packages/grafana-ui/src/components/Graph/Graph.tsx [hunk 13]
```diff
@@ -4,41 +4,66 @@ import React, { PureComponent } from 'react';
...
   timeRange: TimeRange; // NOTE: we should aim to make `time` a property of the axis, not force it for all graphs
+  timeZone: TimeZone; // NOTE: we should aim to make `time` a property of the axis, not force it for all graphs
   showLines?: boolean;
   showPoints?: boolean;
   showBars?: boolean;
   width: number;
   height: number;
+  isStacked?: boolean;
+  lineWidth?: number;
+  onSelectionChanged?: (range: AbsoluteTimeRange) => void;
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Graph/Graph.tsx]: Thinking of making this a bit more generic: Maybe OnHorizontalRegionSelected , can we try to make range a bit more untyped so it could work for non-time y-axis Graphs (so just start, end, or startX, endX)

**Review by torkelo — APPROVED** [review]
Tested this and works great. found 1 little bug with the legend color picker being enabled (but not working). fixed that while I was reviewing it. 

Added some suggestions & ideas for changes.

File: packages/grafana-ui/src/components/Graph/Graph.tsx (hunk 13 above)
Comment (hugohaggmark) [reply packages/grafana-ui/src/components/Graph/Graph.tsx]: Added to #18476

File: packages/grafana-data/src/utils/processDataFrame.ts (hunk 2 above)
Comment (hugohaggmark) [reply packages/grafana-data/src/utils/processDataFrame.ts]: Created this issue to be worked on later: #18528
